srml/staking/src/inflation.rs

```diff
@@ -0,0 +1,327 @@
+		let mut points = vec![];
+
+		// Points for left part
+		points.push((0.0, I_0));
+		points.push((0.5, I_left(0.5)));
+
+		// Approximation for right part.
+		//
+		// We start from 0.5 (x0) and we try to find the next point (x1) for which the linear
+		// approximation of (x0, x1) doesn't deviate from float definition by an error of
+		// GEN_ERROR.
+
+		// When computing deviation between linear approximation and float definition we iterate
+		// over all points with this step precision.
+		const STEP_PRECISION: f64 = 0.000_000_1;
+		// Max error used for generating points.
+		const GEN_ERROR: f64 = 0.000_1;
+
+		let mut x0: f64 = 0.5;
+		let mut x1: f64 = x0;
+
+		// This is just a step used to find next x1:
+		// if x1 + step result in a not enought precise approximation we reduce step and try again.
+		// we stop as soon as step is less than STEP_PRECISION.
+		let mut step: f64 = 0.1;
+
+		loop {
+			let next_x1 = x1 + step;
+
+			if next_x1 >= 1.0 {
+				points.push((1.0, I_right(1.0)));
+				break;
+			}
+
+			let y0 = I_right(x0);
+			let next_y1 = I_right(next_x1);
+
+			let mut error_overflowed = false;
+
+			// Test error is not overflowed
+
+			// Quick test on one point
+			if (I_right((x0 + next_x1)/2.0) - (y0 + next_y1)/2.0).abs() > GEN_ERROR {
+				error_overflowed = true;
+			}
+
+			// Long test on all points
+			if !error_overflowed {
+				let linear = LinearFloat::new(x0, y0, next_x1, next_y1);
+				let mut cursor = x0;
+				while cursor < x1 {
+					if (I_right(cursor) - linear.compute(cursor)).abs() > GEN_ERROR {
+						error_overflowed = true;
+					}
+					cursor += STEP_PRECISION;
+				}
+			}
+
+			if error_overflowed {
+				if step <= STEP_PRECISION {
+					points.push((x1, I_right(x1)));
+					x0 = x1;
+					step = 0.1;
+				} else {
+					step /= 10.0;
+				}
+			} else {
+				x1 = next_x1;
+			}
+		}
+
+		// Convert points to piecewise linear definition
+		let pieces: Vec<(u32, super::Linear)> = (0..points.len()-1)
+			.map(|i| {
+				let p0 = points[i];
+				let p1 = points[i+1];
+
+				let linear = LinearFloat::new(p0.0, p0.1, p1.0, p1.1);
+
+				// Needed if we want to use a Perbill later
+				assert!(linear.a.abs() <= 1.0);
+				// Needed if we want to use a Perbill later
+				assert!(linear.b.abs() <= 1.0);
+				// Needed to stick with our restrictive definition of linear
+				assert!(linear.b.signum() == 1.0);
+
+				(
+					(p0.0 * 1_000_000_000.0) as u32,
+					super::Linear {
+						negative_a: linear.a.signum() < 0.0,
+						a: (linear.a.abs() * 1_000_000_000.0) as u32,
+						b: (linear.b.abs() * 1_000_000_000.0) as u32,
+					}
+				)
+			})
+			.collect();
+
+		println!("Generated pieces: {:?}", pieces);
+		assert_eq!(pieces.len(), 20);
+
+		super::PiecewiseLinear { pieces: (&pieces[..]).try_into().unwrap() }
+	}
+
+	/// This test is only useful to generate a new set of points for the definition of I_NPoS.
+	#[test]
+	fn generate_I_NPOS() {
+		assert_eq!(super::I_NPOS, I_NPoS_points());
+	}
```

Comment:
this test is very long if compiled in debug mode, maybe it worth not executing it, all in all if current definition is precise enough that is fine, this test is just for generating a new one.